### DISCUSSION TITLE
One-to-one existence check on the optional side

### DISCUSSION THREAD

**Discussion by ssukienn** [prisma/prisma#18119] [Q&A]
Hi, I wonder if s1 could help me understand why the existence check is only possible from one side in one-to-one relation.

schema:
```prisma

datasource db {
  provider = "postgresql"
  url      = env("...")
}

generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["clientExtensions"]
}

model Visualisation {
  id                   String               @id @default(uuid())
  staticData     StaticData?
}

model StaticData {
  id                      String            @id @default(uuid())
  visualisation    Visualisation @relation(fields: [visualisationId], references: [id], onDelete: Cascade)
  visualisationId String @unique
}
```

Why this is not possible:
```ts
this.prisma.visualisation.findFirstOrThrow({
        where: { id: visualisationId, staticData: { isNot: null } }, // error on isNot
        include: { staticData: true },
      });
```

I mean it is not possible due to that signature lacking `null` union:
```ts
export type StaticDataRelationFilter = {
  is?: StaticDataWhereInput
  isNot?: StaticDataWhereInput
}
```
but not sure why it is generated like that.

This is possible:
```ts
this.prisma.staticData.findFirstOrThrow({
        where: { id: staticDataId, visualisation: { isNot: null } },
      });
```

In the end, I resorted to:
```ts
this.prisma.staticData.findUniqueOrThrow({
        where: {
          visualisationId,
        },
      });
```
which in fact is a better query in my case.

But I wonder why I can't check for relation existence from the `visualization` model where `staticData` might or might not exist on it? Is it because that `is/isNot` works exclusively based on the presence of FK and that side (`Visualisation`) has no FK? My thinking was that because it is 1:1 it would be possible to determine if `StaticData` is present on `Visualization` with reverse query. Should I assume Prisma is not doing that here?

**Comment by nurul3101** [comment]
Hello @ssukienn 👋 

The visualisation `findFirstOrThrow` query you shared is working as expected for me.

```typescript
import { PrismaClient } from "@prisma/client";

const prisma = new PrismaClient();

// A `main` function so that you can use async/await
async function main() {
  await prisma.visualisation.create({
    data: {
      id: "1",
      staticData: {
        create: {
          id: "1",
        },
      },
    },
  });

  const visualisation = await prisma.visualisation.findFirstOrThrow({
    where: {
      id: "1",
      staticData: {
        isNot: null,
      },
    },
    include: {
      staticData: true,
    },
  });

  console.log("visualisation", visualisation);
}

main()
  .catch((e) => {
    throw e;
  })
  .finally(async () => {
    await prisma.$disconnect();
  });
```

Here's the response:

<img width="619" alt="Screenshot 2023-02-28 at 12 14 38 PM" src="https://user-images.githubusercontent.com/13049676/221838012-2f159ae0-6b1a-4824-b2e9-08ac58a89cfb.png">

**Comment by ssukienn** [comment]
Hello! Thanks for the response 🙏 

I managed to get to some minimal schema that is causing this  (prev one was reduced too much which caused different type generation):

```prisma
model Visualisation {
  id                   String               @id @default(uuid())
  staticData           StaticData?

  >>>configuredProperties ConfiguredProperty[]
}

model StaticData {
  id              String        @id @default(uuid())
  visualisation   Visualisation @relation(fields: [visualisationId], references: [id], onDelete: Cascade)
  visualisationId String        @unique

  properties      Property[]
}

model Property {
  id           String     @id @default(uuid())
  staticData   StaticData @relation(fields: [staticDataId], references: [id], onDelete: Cascade)
  staticDataId String

  configuredProperties ConfiguredProperty[]
}

model ConfiguredProperty {
  id                  String               @id @default(uuid())
  >>>visualisation       Visualisation        @relation(fields: [visualisationId], references: [id], onDelete: Cascade)
  >>>visualisationId     String
  property            Property             @relation(fields: [propertyId], references: [id], onDelete: Cascade)
  propertyId          String
}
```

remove `>>>` and you will face the error I mentioned, comment out those lines, and the error will disappear. 

`visualisation` on `ConfiguredProperty` might seem redundant if you want to maintain referential integrity between `Vis-SData-Prop-CProp` (this design does not enforce `CofiguredProperty` to have FK to the same `Visualization` as one derived deterministically from `Property` FK, through `SData` but it is my responsibility to take care of that in the app-layer) but this FK on that entity is handy for my use case. 

Let's envision a situation where those lines are in place (error case)

This schema tells us that:
* you can have any arbitral `Visualization` entities
  * you are enforced to make `StaticData?` relation side optional (one-to-one) by Prisma
  * all below-mentioned entities might never exists
* if `StaticData` exists it must refer to some unique `Visualisation` (one-to-one)
* if `Property` exists it must refer to some `StaticData` (many-to-one)
* if `ConfiguredProperty` exists it must refer to some `Property` (many-to-one) and to some `Visualisation` (many-to-one) [possibly different Visualization than one from `Property->StaticData->Vis`]

Is it not a flaw then that `StaticDataRelationFilter` does not permit `null` for it's properties (`is/isNot`) on `Visualization` model?

Do I miss something?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
